lib/matplotlib/figure.py

```diff
@@ -1999,7 +1999,7 @@ def __setstate__(self, state):
 
         # re-initialise some of the unstored state information
         self._axobservers = []
-        self.canvas = None
+        FigureCanvasBase(self)  # Set self.canvas.
```

Comment:
I guess it would be vaguely nicer to set state["canvas"] (and also state["_axobservers"], state["_layoutbox"]) to the right values in `__getstate__` rather than popping them out of the dict and having to restore them here, but I'm not going to insist on that.

Reply:
The types in these values tend to not be things that pickle well.  `canvas` can be a Qt object and the observers can be lambdas (and yes, we could use cloud pickle and go down the route that dask does to pickle/unpickle almost anything, but that is probably not worth the complexity or something we want to encourage).

The other nice thing about not directly pickling the canvas objects is that when re-hydrated they will come up in the correct framework of the importer.  This fix only matters if you are unpickling a `Figure` that was unknown to `pyplot` when it was pickled (and `inline` de-registers the figures after they show) then this code path will matter.

From a philosophical I think that the canvas object is not part of the identity of the `Figure`.  A given `Figure` can only we associated with one canvas at a time (due to the circular references) but we can (and do!) switch the canvas out "live", hence even if we could, it probably should not be included verbatim in the pickle.

Reply:
Yes, I know you can't pickle Qt canvasses (and agree they're not really part of figure state); I meant to pickle a FigureCanvasBase instead (I guess you only need to do so if restore_to_pylab is false).  Not a really big deal, though.